### ISSUE TITLE
Docker slow to respond

### ISSUE THREAD

**Issue by skiwithpete** [cointop-sh/cointop#265]
(thanks for updating the docker to .10)

Actions in Docker are taking a very long time to render.  For example, smashing the [ key stops showing the change progress (from 3d to 7d for example) after a couple of presses and can take up to 10seconds to update.  Then it takes an age to update the graph as well.

Another example is Shift+J or Shift+K taking 4 or 5 seconds for the graph to resize (whereas in earlier releases it was instantaneous).

I'm using on a 2019 i3 server, with a 2019 macbook pro in terminal. 

Let me know what additional info I can provide to help find the cause.

Thanks,

Pete

**Comment by lyricnz** [comment]
I am not experiencing this (docker on macbook pro).  How are you running docker? 

https://user-images.githubusercontent.com/122371/140664892-89d5243f-2122-48e8-981c-8bede7e812fa.mov

**Comment by lyricnz** [comment]
I actually made a change recently #256  to improve the refresh after changing currencies. Seems quick here?  

https://user-images.githubusercontent.com/122371/140664948-c3cf40d2-53ca-4311-b5ca-f4f72c254e32.mov

**Comment by skiwithpete** [comment]
Actually, it happens in portfolio view.  And then, when I leave Portfolio view, it affects the main page as well.

**Comment by lyricnz** [comment]
I can see a problem with swapping back and forth to portfolio mode (at least since previous release). Raised #266 on that.  But doesn't seem to be _slow_.

**Comment by skiwithpete** [comment]
In portfolio mode pressing [ or ] doesn't go slow for you?

I'm running both a pretty powerful server and a powerful laptop... 

Might be worth messing around with quantities of coins in your portfolio to see the problem I'm having.

**Comment by lyricnz** [comment]
Yes, [ and ] are slow when you have lots of coins.   It looks like it's fetching the history for each coin, one by one (with a 2 second delay between, to avoid coingecko rate limiting etc).  Wasn't that always the case?

```
$ DEBUG=1 DEBUG_HTTP=1 ./bin/cointop
....
$ grep doReq /tmp/cointop.log  | grep days=7
time="2021-11-08T12:59:57+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/wrapped-bitcoin/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:00+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/bitcoin/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:02+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/ethereum/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:04+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/binancecoin/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:07+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/solana/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:10+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/litecoin/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:12+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/avalanche-2/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:14+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/polkadot/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:17+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/avaterra/market_chart?days=7&vs_currency=AUD"
time="2021-11-08T13:00:19+11:00" level=debug msg="doReq GET https://api.coingecko.com/api/v3/coins/binance-peg-xrp/market_chart?days=7&vs_currency=AUD"
```

**Comment by lyricnz** [comment]
This delay has been in place for 3 years
https://github.com/cointop-sh/cointop/blame/master/cointop/chart.go#L255

**Comment by skiwithpete** [comment]
Then is it an ordering bug?  Like it should prioritize the change of state
with the [ and cease the previous operation to avoid the slowdown?

**Comment by lyricnz** [comment]
It's not doing anything else at the time, from what I can tell.  The 2-second delay is intended to avoid rate-limiting. It would be very easy to hit 50 requests/minute with a couple of range-changes across a large portfolio.  There's an existing issue to investigate/tune/minimize this #228

**Comment by lyricnz** [comment]
Maaaaaybe it would be possible to "fake" the new time range data (using the previous range, and dropping ~half the data), and update it in the background?  Especially if moving to a shorter time range.
